### PR TITLE
feat: subcontract code refactor and enhancement

### PR DESCRIPTION
**Pull request by rohitwaghchaure**
Done code refactoring and introduced new class Subcontracting in new file subcontracting.py

Feature Enhancement

1. Provision to edit consumed qty in the Purchase Receipt
Screenshot 2021-05-29 at 3 44 29 PM

2. Provision to get non consumed components from the Supplier
Screenshot 2021-05-24 at 6 42 48 PM

**Added Subcontract Order Summary Report** 

image

Docs https://github.com/frappe/erpnext_documentation/pull/343

### CODE REVIEW

**Comment by rohitwaghchaure** [comment]
Dependency https://github.com/frappe/frappe/pull/13439

File: erpnext/buying/doctype/purchase_order/purchase_order.js [hunk 1]
```diff
@@ -45,6 +45,38 @@ frappe.ui.form.on("Purchase Order", {
...
+				}
+			});
+		}
+
+		if (po_details && po_details.length) {
+			frm.add_custom_button(__('Return of Components'), () => {
+				frm.call({
+					method: 'erpnext.buying.doctype.purchase_order.purchase_order.get_materials_from_supplier',
+					freeze_message: __('Creating Stock Entry'),
+					args: { purchase_order: frm.doc.name, po_details: po_details },
```
Comment (marination) [inline erpnext/buying/doctype/purchase_order/purchase_order.js]:
```suggestion
					method: 'erpnext.buying.doctype.purchase_order.purchase_order.get_materials_from_supplier',
					freeze: true,
					freeze_message: __('Creating Stock Entry'),
					args: { purchase_order: frm.doc.name, po_details: po_details },
```

File: erpnext/buying/doctype/purchase_order/purchase_order.js [hunk 2]
```diff
@@ -504,12 +536,14 @@ erpnext.buying.PurchaseOrderController = erpnext.buying.BuyingController.extend(
 			],
 			primary_action: function() {
 				var data = d.get_values();
+				var content_msg = 'Reason for hold: ' + data.reason_for_hold;
```
Comment (marination) [inline erpnext/buying/doctype/purchase_order/purchase_order.js]:
```suggestion
				let reason_for_hold = 'Reason for hold: ' + data.reason_for_hold;
```

File: erpnext/buying/doctype/purchase_order/purchase_order.js [hunk 3]
```diff
@@ -504,12 +536,14 @@ erpnext.buying.PurchaseOrderController = erpnext.buying.BuyingController.extend(
...
+
 				frappe.call({
 					method: "frappe.desk.form.utils.add_comment",
 					args: {
 						reference_doctype: me.frm.doctype,
 						reference_name: me.frm.docname,
-						content: __('Reason for hold: ')+data.reason_for_hold,
+						content: __(content_msg),
```
Comment (marination) [inline erpnext/buying/doctype/purchase_order/purchase_order.js]:
```suggestion
						content: __(reason_for_hold),
```

File: erpnext/buying/doctype/purchase_order/purchase_order.py [hunk 4]
```diff
@@ -468,9 +468,10 @@ def update_item(obj, target, source_parent):
 
 @frappe.whitelist()
 def make_rm_stock_entry(purchase_order, rm_items):
+	rm_items_list = rm_items
 	if isinstance(rm_items, string_types):
```
Comment (marination) [inline erpnext/buying/doctype/purchase_order/purchase_order.py]:
```suggestion
	if isinstance(rm_items, str):
```
`string_types` from `six` returns `str` only in py3. Let's remove it's usage wherever possible
Screenshot 2021-06-08 at 1 17 42 AM

File: erpnext/buying/doctype/purchase_order/purchase_order.py [hunk 5]
```diff
@@ -547,3 +550,57 @@ def update_status(status, name):
 def make_inter_company_sales_order(source_name, target_doc=None):
 	from erpnext.accounts.doctype.sales_invoice.sales_invoice import make_inter_company_transaction
 	return make_inter_company_transaction("Purchase Order", source_name, target_doc)
+
+@frappe.whitelist()
+def get_materials_from_supplier(purchase_order, po_details):
+	if isinstance(po_details, string_types):
```
Comment (marination) [inline erpnext/buying/doctype/purchase_order/purchase_order.py]:
```suggestion
	if isinstance(po_details, str):
```

File: erpnext/buying/report/subcontract_order_summary/subcontract_order_summary.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+			data.append(res)
+
+def get_columns():
+	return [
+		{
+			"label": _("Id"),
+			"fieldname": "po_id",
+			"fieldtype": "Link",
+			"options": "Purchase Order",
+			"width": 100
```
Comment (marination) [inline erpnext/buying/report/subcontract_order_summary/subcontract_order_summary.py]: `Order No.` or `Purchase Order`
```suggestion
			"label": _("Order No."),
			"fieldname": "po_id",
			"fieldtype": "Link",
			"options": "Purchase Order",
			"width": 200
```

File: erpnext/controllers/subcontracting.py [hunk 7]
```diff
@@ -0,0 +1,342 @@
+from __future__ import unicode_literals
```
Comment (marination) [inline erpnext/controllers/subcontracting.py]: Not needed for py3
```suggestion
```

File: erpnext/controllers/subcontracting.py [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+
+import frappe
+from frappe import _
+from frappe.utils import flt, cint
+from collections import defaultdict
+from erpnext.stock.doctype.serial_no.serial_no import get_serial_nos
+
+class Subcontracting(object):
```
Comment (marination) [inline erpnext/controllers/subcontracting.py]:
```suggestion
class Subcontracting():
```
Passing `object` is [not needed in py3](https://stackoverflow.com/questions/7375595/class-with-object-as-a-parameter) as it was in py2

File: erpnext/controllers/subcontracting.py [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+		self.set_supplied_items()
+
+	def initialized_fields(self):
+		self.available_materials = frappe._dict()
+		self.alternative_item_details = frappe._dict()
+		self.get_backflush_based_on()
+
+	def get_backflush_based_on(self):
+		self.backflush_based_on = frappe.db.get_single_value("Buying Settings",
+			"backflush_raw_materials_of_subcontract_based_on")
```
Comment (marination) [inline erpnext/controllers/subcontracting.py]: Could add this in class `init` so it doesn't have to be explicitly invoked
```py
def __init__(self):
    self.backflush_based_on = frappe.db.get_single_value("Buying Settings", 
        "backflush_raw_materials_of_subcontract_based_on")

def set_materials_for_subcontracted_items(self, raw_material_table):
    ...
```

File: erpnext/stock/doctype/stock_entry/stock_entry.py [hunk 10]
```diff
@@ -1400,33 +1398,26 @@ def validate_batch(self):
...
-			item_code = d.get('original_item') or d.get('item_code')
-			reserve_warehouse = item_wh.get(item_code)
-			stock_bin = get_bin(item_code, reserve_warehouse)
-			stock_bin.update_reserved_qty_for_sub_contracting()
+			#Get PO Supplied Items Details
+			item_wh = frappe._dict(frappe.db.sql("""
+				select rm_item_code, reserve_warehouse
+				from `tabPurchase Order` po, `tabPurchase Order Item Supplied` poitemsup
+				where po.name = poitemsup.parent
+				and po.name = %s""", self.purchase_order))
```
Comment (marination) [inline erpnext/stock/doctype/stock_entry/stock_entry.py:1411]:
```suggestion
				select 
				    rm_item_code, reserve_warehouse
				from 
				    `tabPurchase Order` po, `tabPurchase Order Item Supplied` poitemsup
				where 
				    po.name = poitemsup.parent
				    and po.name = %s""", self.purchase_order))
```

**Review by marination — CHANGES_REQUESTED** [review]

File: erpnext/controllers/subcontracting.py (hunk 9 above)
Comment (rohitwaghchaure) [reply erpnext/controllers/subcontracting.py]: __init__ will call every time even though the order is not subcontracted

**Comment by rohitwaghchaure** [comment]
image

File: erpnext/controllers/subcontracting.py [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+		if self.backflush_based_on != 'BOM' and flt(row.consumed_qty) == 0.0:
+			msg = f'Row {row.idx}: the consumed qty cannot be zero for the item {frappe.bold(row.rm_item_code)}'
+
+			frappe.throw(_(msg),title=_('Consumed Items Qty Check'))
+
+	def __validate_batch_no(self, row, key):
+		if row.get('batch_no') and row.get('batch_no') not in self.__transferred_items.get(key).get('batch_no'):
+			link = get_link_to_form('Purchase Order', row.purchase_order)
+			msg = f'The Batch No {frappe.bold(row.get("batch_no"))} has not supplied against the Purchase Order {link}'
+			frappe.throw(_(msg))
```
Comment (marination) [inline erpnext/controllers/subcontracting.py]:
```suggestion
			msg = f'The Batch No {frappe.bold(row.get("batch_no"))} was not supplied against the Purchase Order {link}'
			frappe.throw(_(msg), title=_("Incorrect Batch Consumed"))
```

File: erpnext/controllers/subcontracting.py [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+	def __validate_serial_no(self, row, key):
+		if row.get('serial_no'):
+			serial_nos = get_serial_nos(row.get('serial_no'))
+			incorrect_sn = set(serial_nos).difference(self.__transferred_items.get(key).get('serial_no'))
+
+			if incorrect_sn:
+				incorrect_sn = "\n".join(incorrect_sn)
+				link = get_link_to_form('Purchase Order', row.purchase_order)
+				msg = f'The Serial Nos {incorrect_sn} has not supplied against the Purchase Order {link}'
+				frappe.throw(_(msg))
```
Comment (marination) [inline erpnext/controllers/subcontracting.py]:
```suggestion
				msg = f'The Serial Nos {incorrect_sn} were not supplied against the Purchase Order {link}'
				frappe.throw(_(msg), title=_("Incorrect Serial No. Consumed"))
```

File: erpnext/controllers/subcontracting.py [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+				self.__set_consumed_qty(rm_obj, qty)
+		else:
+			self.__set_consumed_qty(rm_obj, qty, bom_item.required_qty or qty)
+			self.__set_serial_nos(item_row, rm_obj)
+
+	def __set_consumed_qty(self, rm_obj, consumed_qty, required_qty=0):
+		rm_obj.required_qty = required_qty
+		rm_obj.consumed_qty = consumed_qty
+
+	def __set_batch_no_as_per_qty(self, item_row, bom_item, rm_obj, batch_no, qty):
```
Comment (marination) [inline erpnext/controllers/subcontracting.py]: `bom_item` unused
```suggestion
	def __set_batch_no_as_per_qty(self, item_row, rm_obj, batch_no, qty):
```

File: erpnext/controllers/subcontracting.py [hunk 14]
```diff
@@ -0,0 +1,393 @@
...
+				if batch_qty >= qty:
+					self.__set_batch_no_as_per_qty(item_row, bom_item, rm_obj, batch_no, qty)
+					self.available_materials[key]['batch_no'][batch_no] -= qty
+					return
+
+				elif qty > 0 and batch_qty > 0:
+					qty -= batch_qty
+					new_rm_obj = self.append(self.raw_material_table, bom_item)
+					new_rm_obj.reference_name = item_row.name
+					self.__set_batch_no_as_per_qty(item_row, bom_item, new_rm_obj, batch_no, batch_qty)
```
Comment (marination) [inline erpnext/controllers/subcontracting.py]: `bom_item` unused
```suggestion
					self.__set_batch_no_as_per_qty(item_row, new_rm_obj, batch_no, batch_qty)
```

File: erpnext/controllers/subcontracting.py [hunk 15]
```diff
@@ -0,0 +1,393 @@
...
+			self.__set_batch_nos(bom_item, item_row, rm_obj, qty)
+
+	def __set_batch_nos(self, bom_item, item_row, rm_obj, qty):
+		key = (rm_obj.rm_item_code, item_row.item_code, item_row.purchase_order)
+
+		if (self.available_materials.get(key) and self.available_materials[key]['batch_no']):
+			new_rm_obj = None
+			for batch_no, batch_qty in self.available_materials[key]['batch_no'].items():
+				if batch_qty >= qty:
+					self.__set_batch_no_as_per_qty(item_row, bom_item, rm_obj, batch_no, qty)
```
Comment (marination) [inline erpnext/controllers/subcontracting.py]: `bom_item` unused
```suggestion
					self.__set_batch_no_as_per_qty(item_row, rm_obj, batch_no, qty)
```

**Review by marination — CHANGES_REQUESTED** [review]

**Review by marination — APPROVED** [review]

**Pull request merged**